In [1]:
from __future__ import absolute_import
from __future__ import print_function
import numpy 
import numpy as np
import os
import keras
from tensorflow.python.keras.layers import Dense
from tensorflow.python.keras import Sequential
from keras.preprocessing import sequence
from keras.models import Sequential, Model
from keras.layers import Dense, Dropout, Bidirectional,Input,Activation,Flatten
from keras.datasets import imdb
from keras import backend as K
from keras.layers import Conv1D, GlobalAveragePooling1D, MaxPooling1D,GlobalMaxPooling1D
import typing
from keras.utils.np_utils import to_categorical
from keras import Model
from keras import layers
from keras import Input
num_classes = 5
epochs = 2


Using TensorFlow backend.


In [2]:
def create_base_network(input_shape):  
 
   input = Input(shape = input_shape)
   x = Conv1D(64,1, activation='relu' )(input)
   #x = Dropout(0.5)(x)
   #x = MaxPooling1D(pool_size=1)(x)
   #x = Conv1D(64,1, activation='relu')(x) 
   #x = Dropout(0.5)(x)
   #x = MaxPooling1D(pool_size=1)(x)
   #x = Conv1D(64,1, activation='relu')(x) 
   return Model(input,x)

dataset = numpy.loadtxt("trx1.csv", delimiter=",",dtype= 'int')
dataset2 = numpy.loadtxt("trx2.csv", delimiter=",",dtype= 'int')
datasety = numpy.loadtxt("datay.csv", delimiter=",",dtype= 'int')
datasete = numpy.loadtxt("tex1.csv", delimiter=",",dtype= 'int')
datasete2 = numpy.loadtxt("tex2.csv", delimiter=",",dtype= 'int')
datasetye = numpy.loadtxt("tey.csv", delimiter=",",dtype= 'int')

A = np.resize(dataset , (224,1,900))
print("x_train",A.shape)
print("x_train[1:]",A.shape[1:])
B = np.resize(dataset2 , (224,1,900))
print(B.shape)
C = np.resize(datasety , (224,1))
print(C.shape)

pair = np.vstack((A,B))   # vertical stack

pair = np.resize(pair , (224,2,1,900))
print("pair",pair.shape)
print("pair[:,0]",pair[:,0].shape)
print("pair[:,1]",pair[:,1].shape)

a = np.resize(datasete , (57,1,900))
print(a.shape)
b = np.resize(datasete2 , (57,1,900))
print(b.shape)
c = np.resize(datasetye , (57,1))
print(c.shape)

paire = np.vstack((a,b))   # vertical stack

paire = np.resize(paire , (57,2,1,900))

print("paire",paire.shape)
print("paire1",paire[:,0].shape)
print("paire2",paire[:,1].shape)

y1= keras.utils.to_categorical(C,num_classes=5)
y2= keras.utils.to_categorical(c,num_classes=5)
y1c = np.resize(y1 , (224,5,1))
y2c = np.resize(y2 , (57,5,1))
print("y1",y1.shape)
kernel_size = 3
hidden_dims = 64
epochs = 20

input_shape = A.shape[1:]
base_network = create_base_network(input_shape)

input_a = Input(shape=input_shape)
input_b = Input(shape=input_shape)

processed_a = base_network(input_a)
processed_b = base_network(input_b)

answer_size = 5

concatenate = layers.concatenate([processed_a, processed_b],axis = -1)
y3 = Flatten()(concatenate)
lasty = layers.Dense(answer_size,
                      activation = 'softmax')(y3)
model = Model([input_a, input_b], lasty) 
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
model.summary()
model.fit([pair[:, 0], pair[:, 1]], y1,
          batch_size=1,
          epochs=epochs,
          validation_data=([paire[:, 0], paire[:, 1]], y2))


x_train (224, 1, 900)
x_train[1:] (1, 900)
(224, 1, 900)
(224, 1)
pair (224, 2, 1, 900)
pair[:,0] (224, 1, 900)
pair[:,1] (224, 1, 900)
(57, 1, 900)
(57, 1, 900)
(57, 1)
paire (57, 2, 1, 900)
paire1 (57, 1, 900)
paire2 (57, 1, 900)
y1 (224, 5)
Model: "model_2"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
input_2 (InputLayer)            (None, 1, 900)       0                                            
__________________________________________________________________________________________________
input_3 (InputLayer)            (None, 1, 900)       0                                            
__________________________________________________________________________________________________
model_1 (Model)                 (None, 1, 64)        57664       input_2[0][0]                    
                                              

x_train (224, 1, 900)
x_train[1:] (1, 900)
(224, 1, 900)
(224, 1)
pair (224, 2, 1, 900)
pair[:,0] (224, 1, 900)
pair[:,1] (224, 1, 900)
(57, 1, 900)
(57, 1, 900)
(57, 1)
paire (57, 2, 1, 900)
paire1 (57, 1, 900)
paire2 (57, 1, 900)
y1 (224, 5)
Model: "model_2"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
==================================================================================================
input_2 (InputLayer)            (None, 1, 900)       0                                            
__________________________________________________________________________________________________
input_3 (InputLayer)            (None, 1, 900)       0                                            
__________________________________________________________________________________________________
model_1 (Model)                 (None, 1, 64)        57664       input_2[0][0]                    
                                                                 input_3[0][0]                    
__________________________________________________________________________________________________
concatenate_1 (Concatenate)     (None, 1, 128)       0           model_1[1][0]                    
                                                                 model_1[2][0]                    
__________________________________________________________________________________________________
flatten_1 (Flatten)             (None, 128)          0           concatenate_1[0][0]              
__________________________________________________________________________________________________
dense_1 (Dense)                 (None, 5)            645         flatten_1[0][0]                  
==================================================================================================
Total params: 58,309
Trainable params: 58,309
Non-trainable params: 0
__________________________________________________________________________________________________
Train on 224 samples, validate on 57 samples
Epoch 1/20
224/224 [==============================] - 7s 31ms/step - loss: 4.4170 - accuracy: 0.2188 - val_loss: 1.7621 - val_accuracy: 0.2281
Epoch 2/20
224/224 [==============================] - 1s 5ms/step - loss: 1.5768 - accuracy: 0.3393 - val_loss: 1.5475 - val_accuracy: 0.2456
Epoch 3/20
224/224 [==============================] - 1s 5ms/step - loss: 1.5431 - accuracy: 0.3527 - val_loss: 1.5519 - val_accuracy: 0.2105
Epoch 4/20
224/224 [==============================] - 1s 5ms/step - loss: 1.4918 - accuracy: 0.3393 - val_loss: 1.5981 - val_accuracy: 0.2807
Epoch 5/20
224/224 [==============================] - 1s 5ms/step - loss: 1.5049 - accuracy: 0.3527 - val_loss: 1.4879 - val_accuracy: 0.2281
Epoch 6/20
224/224 [==============================] - 1s 5ms/step - loss: 1.4105 - accuracy: 0.3571 - val_loss: 1.4990 - val_accuracy: 0.2982
Epoch 7/20
224/224 [==============================] - 1s 5ms/step - loss: 1.3849 - accuracy: 0.3527 - val_loss: 1.5634 - val_accuracy: 0.2632
Epoch 8/20
224/224 [==============================] - 1s 5ms/step - loss: 1.3903 - accuracy: 0.3571 - val_loss: 1.5206 - val_accuracy: 0.2632
Epoch 9/20
224/224 [==============================] - 1s 5ms/step - loss: 1.3925 - accuracy: 0.3705 - val_loss: 1.5087 - val_accuracy: 0.2632
Epoch 10/20
224/224 [==============================] - 1s 5ms/step - loss: 1.3588 - accuracy: 0.3795 - val_loss: 1.5243 - val_accuracy: 0.2105
Epoch 11/20
224/224 [==============================] - 1s 5ms/step - loss: 1.3453 - accuracy: 0.3929 - val_loss: 1.6731 - val_accuracy: 0.2632
Epoch 12/20
224/224 [==============================] - 1s 5ms/step - loss: 1.3315 - accuracy: 0.3929 - val_loss: 1.5655 - val_accuracy: 0.2281
Epoch 13/20
224/224 [==============================] - 1s 5ms/step - loss: 1.3226 - accuracy: 0.4018 - val_loss: 1.8017 - val_accuracy: 0.2456
Epoch 14/20
224/224 [==============================] - 1s 5ms/step - loss: 1.3091 - accuracy: 0.3839 - val_loss: 1.6262 - val_accuracy: 0.2632
Epoch 15/20
224/224 [==============================] - 1s 5ms/step - loss: 1.3192 - accuracy: 0.4018 - val_loss: 1.7419 - val_accuracy: 0.2632
Epoch 16/20
224/224 [==============================] - 1s 5ms/step - loss: 1.3271 - accuracy: 0.3705 - val_loss: 1.6180 - val_accuracy: 0.2456
Epoch 17/20
224/224 [==============================] - 1s 5ms/step - loss: 1.2991 - accuracy: 0.4062 - val_loss: 1.6343 - val_accuracy: 0.2105
Epoch 18/20
224/224 [==============================] - 1s 5ms/step - loss: 1.2700 - accuracy: 0.4107 - val_loss: 1.6592 - val_accuracy: 0.2456
Epoch 19/20
224/224 [==============================] - 1s 5ms/step - loss: 1.2783 - accuracy: 0.4196 - val_loss: 2.1773 - val_accuracy: 0.2281
Epoch 20/20
224/224 [==============================] - 1s 5ms/step - loss: 1.2988 - accuracy: 0.4062 - val_loss: 1.5986 - val_accuracy: 0.2632
<keras.callbacks.callbacks.History at 0x7fec1a8a6e10>

In [0]:
import numpy as np
dataset = numpy.loadtxt("trx1.csv", delimiter=",",dtype= 'int')
dataset2 = numpy.loadtxt("trx2.csv", delimiter=",",dtype= 'int')
datasety = numpy.loadtxt("datay.csv", delimiter=",",dtype= 'int')
datasete = numpy.loadtxt("tex1.csv", delimiter=",",dtype= 'int')
datasete2 = numpy.loadtxt("tex2.csv", delimiter=",",dtype= 'int')
datasetye = numpy.loadtxt("tey.csv", delimiter=",",dtype= 'int')
len(dataset)#224
np.size(dataset)#201600
np. size(dataset,1)#900
np. shape (dataset)#(224, 900)


(224, 900)